src/main/java/org/veupathdb/service/eda/compute/controller/ComputeController.java

```diff
@@ -0,0 +1,201 @@
+   */
+  private <R extends ComputeRequestBase, C> JobResponse submitJob(PluginProvider<R, C> plugin, R entity) {
+    var auth = UserProvider.getSubmittedAuth(request).orElseThrow();
+
+    requirePermissions(entity, auth);
+
+    // Validate the request body.
+    plugin.getValidator()
+      .validate(entity);
+
+    return EDA.getOrSubmitComputeJob(plugin, entity, auth);
+  }
+
+  /**
+   * Returns a stream over the contents of the target {@code file} from the job
+   * described by the given configuration {@code entity}.
+   *
+   * @param plugin Metadata about the plugin to which the target job belongs.
+   *
+   * @param file Name of the filetype target.
+   * <p>
+   * One of "meta", "tabular", or "statistics".
+   *
+   * @param entity Job configuration request body.
+   *
+   * @param responseFn Controller response type method.
+   *
+   * @return The controller response type generated by the given
+   * {@code responseFn} function.
+   *
+   * @param <R> Type of response the controller response method returns.
+   *
+   * @param <P> Type of the job configuration request body.
+   */
+  public <R extends ResponseDelegate, P extends ComputeRequestBase> R resultFile(
+    PluginMeta<P> plugin,
+    String file,
+    ComputeRequestBase entity,
+    Function<Object, R> responseFn
+  ) {
+    // If there was no plugin with the given name, throw a 404
+    if (plugin == null)
+      throw new NotFoundException();
+
+    requirePermissions(entity, null);
+
+    var jobFiles = EDA.getComputeJobFiles(plugin, entity);
+
+    var fileName = switch(file) {
+      case "meta"       -> ReservedFiles.OutputMeta;
+      case "tabular"    -> ReservedFiles.OutputTabular;
+      case "statistics" -> ReservedFiles.OutputStats;
+      default           -> throw new NotFoundException();
+    };
+
+    var fileRef = jobFiles.stream()
+      .filter(f -> f.getName().equals(fileName))
+      .findFirst()
+      .orElseThrow(NotFoundException::new);
+
+    return responseFn.apply((StreamingOutput) output -> {
+      try(var input = fileRef.open()) {
+        input.transferTo(output);
+      }
+    });
+  }
+
+  /**
+   * Ensures that the requester has the required permission(s) to call a plugin
+   * endpoint for a target study.
+   * <p>
+   * As of the time of this writing, the permission required for plugin
+   * execution and lookup is "allowVisualizations".
+   *
+   * @param entity Raw request body containing the ID of the study the user must
+   * have permissions on.
+   *
+   * @param auth The auth header to use in validation, or {@code null} for the
+   * auth header from the {@link #request} context to be used.
+   *
+   * @throws ForbiddenException If the requester does not have the required
+   * permission(s) on the target study.
+   */
+  private void requirePermissions(@NotNull ComputeRequestBase entity, @Nullable Tuples.TwoTuple<String, String> auth) {
+    if (auth == null)
+      auth = UserProvider.getSubmittedAuth(request).orElseThrow();
```

Comment:
what exception is being thrown here?  why not make it explicit and include a message.